File: 1101-1200/1136-Parallel Courses/1136-Parallel Courses.py
import collections
class Solution:
    def minimumSemesters(self, N: int, relations: List[List[int]]) -> int:
        graph = collections.defaultdict(list)
        degrees = [0] * (N + 1)
        for x, y in relations:
            graph[x].append(y)
            degrees[y] += 1
            
        Q = collections.deque([i for i in range(1, N + 1) if degrees[i] == 0])
        course = semester = 0
        while Q:
            semester += 1
            for _ in range(len(Q)):
                curr = Q.popleft()
                course += 1
                for n in graph[curr]:
                    degrees[n] -= 1
                    if degrees[n] == 0:
                        Q.append(n)
        return semester if course == N else -1

File: 1101-1200/1136-Parallel Courses/1136-Parallel Courses.cpp
class Solution {
public:
    int minimumSemesters(int N, vector<vector<int>>& relations) {
        vector<vector<int>> graph(1 + N);
        vector<int> degrees(1 + N);
        for (auto r : relations) {
            graph[r[0]].push_back(r[1]);
            ++degrees[r[1]];
        }
        
        queue<int> Q;
        for (int i = 1; i <= N; ++i) {
            if (degrees[i] == 0) {
                Q.push(i);
            }
        }
        int course = 0, semester = 0;
        while (!Q.empty()) {
            ++semester;
            int qs = Q.size();
            for (int i = 0; i < qs; ++i) {
                int curr = Q.front();
                Q.pop();
                ++course;
                for (int n : graph[curr]) {
                    if (--degrees[n] == 0) {
                        Q.push(n);
                    }
                }
            }
        }
        
        return course == N ? semester : -1;
    }
};
